_posts/2024-04-15-heapdump-oome-native.adoc

```diff
@@ -0,0 +1,258 @@
+---
+layout: post
+title: 'Obtaining heap dump on OutOfMemoryError with Quarkus native'
+date: 2024-04-15
+tags: native heapdump oome epsilongc jol
+synopsis: 'Generating and inspecting heap dumps in Quarkus native when `OutOfMemoryError` errors occur'
+author: galderz
+---
+
+Starting with GraalVM for JDK 21,
+native executables can run with the `-XX:+HeapDumpOnOutOfMemoryError` option to generate a heap dump when a `java.lang.OutOfMemoryError` is thrown.
+In this blog post we will explore how to use the flag, we will inspect what a native image heap dump looks like and how it compares with one produced by HotSpot.
+
+To see this flag in action, we need to manufacture a situation where a Quarkus application runs out of memory.
+One easy way to achieve this is to configure the application with a garbage collector that doesn't do any memory reclamation,
+and that's exactly what the Epsilon GC does.
+Once the Quarkus application is running with Epsilon GC,
+apply some load and within a short space of time it will run out of memory and produce a heap dump.
+
+Let’s start by producing a heap dump with a Quarkus native executable.
+A good starting point would be a Quarkus application that simply responds to an HTTP endpoint request, e.g. the
+https://github.com/quarkusio/quarkus-quickstarts/tree/main/getting-started-reactive[getting-started-reactive] Quarkus quickstart.
```

Comment:
Perhaps it's simpler to use https://code.quarkus.io/

E.g. I can follow this post with:

```
wget "https://code.quarkus.io/d?S=io.quarkus.platform%3A3.8&cn=code.quarkus.io" -O code.zip
unzip code.zip
cd code-with-quarkus
```

Note I explicitly set the version to Quarkus 3.8 to hopefully keep the instructions working for some time.